### PR TITLE
[Bugfix][MoE] Warn when gate fusion silently bypasses the gate's forward hooks

### PR DESCRIPTION
**Pull request by Mohammad-Nabulsi**
## Purpose

Registering a `forward_hook` on a MoE layer's gate module is the obvious way to observe router logits. When shared-expert gate fusion is active, **it installs cleanly, raises nothing, and captures nothing.**

The cause is in `MoERunner`:

```python
self._fse_fuse_gate = gate is not None and shared_expert_gate is not None
```

and later, in the forward:

```python
if self.gate is not None:
    if self._fse_fuse_gate:
        self._maybe_fuse_gate_weights()
        router_logits = F.linear(hidden_states, self._combined_gate_weight)
    else:
        router_logits, _ = self.gate(hidden_states)
```

Under fusion the gate's weights are folded into a combined matrix and the logits are computed with a direct `F.linear`. The `gate` module object still exists and the hook is still attached to it, but the module is never called.

This is a bad failure mode: no exception, no warning, no log line, and an empty capture that is indistinguishable from a layer that genuinely routes nothing. Anyone instrumenting MoE routing — for expert-utilisation studies, router-skew analysis, or debugging a placement policy — tries this first, and nothing tells them it cannot work. Models that fuse shared experts hit it by default.

## Proposed change

Warn once per layer, at forward time, when gate fusion is active *and* the gate module carries forward or forward-pre hooks, and point at the seam that does see the full `router_logits` on every path.

Checked at forward rather than at construction on purpose: instrumentation is normally attached after the model is built, so a constructor-time check would miss the common case.

The warning is only emitted when hooks are actually registered, so it is silent for everyone who is not instrumenting. It uses `logger.warning_once` and names the layer via the already-plumbed `self.layer_name`.

## Test Plan

New: `tests/kernels/moe/test_gate_hook_bypass_warning.py`

```
pytest tests/kernels/moe/test_gate_hook_bypass_warning.py
```

CPU-only, no GPU required. Four cases: a registered forward hook warns and names the layer; a forward-pre hook warns; no hooks stays silent; no gate stays silent.

## Test Result

The four branch outcomes were verified by running the helper's logic against stub modules:

```
  hook     -> warned=True
  prehook  -> warned=True
  none     -> warned=False
  nogate   -> warned=False
```

`ruff check` and `ruff format --check` are clean on both changed files.

**Being straight about it:** I do not have a working vLLM build on the machine this was written on, so I have not executed the pytest file itself — the result above came from running the branch logic standalone against stubs. The tests use `MoERunner.__new__` to avoid constructing a full runner, so they should be cheap in CI, but please run them before trusting them.

I am happy to take this further if maintainers prefer a different shape — for example raising instead of warning, or exposing a supported capture seam rather than only warning about the unsupported one. Warning seemed like the smallest change that removes the silent failure.

### CODE REVIEW

**Comment by Mohammad-Nabulsi** [comment]
Pushed a revision. Two things I got wrong in the first version, both found by reading vLLM's own logging setup more carefully:

1. **The check ran on every forward.** It now runs from the one-shot `_maybe_fuse_gate_weights()` first-call branch instead, so it costs nothing per forward and cannot interact with `torch.compile` tracing of the hot path. It still runs after weight loading, which is the point — instrumentation is normally attached after module construction, so a constructor-time check would miss the common case.

2. **The tests asserted on `caplog`, which could never have passed.** The `vllm` logger is configured with `propagate: False`, so `caplog` never sees it. Separately, `warning_once` is `@lru_cache`d, so the second test's identical message would have been dropped process-wide regardless of ordering. The tests now record `warning_once` calls directly via `monkeypatch`, which fixes both and makes them order-independent.

I should also be explicit about a limitation: I do not have a working vLLM build available, so I have not executed these tests. The logic was verified standalone and the file is ruff-clean, but CI is the first real run. Happy to iterate on anything it turns up.
